src/Channels.Networking.TLS/SecureChannel.cs

```diff
@@ -0,0 +1,132 @@
+                if (token != null)
+                {
+                    var writeBuff = _lowerOutputChannel.Alloc(token.Length);
+                    writeBuff.Write(token);
+                    await writeBuff.FlushAsync();
+                }
+
+                while (true)
+                {
+                    
+
+                    var buffer = await _lowerInputChannel.ReadAsync();
+
+                    ReadCursor pointToSliceMessage;
+                    var f = buffer.CheckForFrameType(out pointToSliceMessage);
+                    while (f != TlsFrameType.Incomplete)
+                    {
+                        if (f == TlsFrameType.Handshake || f == TlsFrameType.ChangeCipherSpec)
+                        {
+                            var messageBuffer = buffer.Slice(0, pointToSliceMessage);
+                            buffer = buffer.Slice(pointToSliceMessage);
+                            var buff = securityContext.ProcessContextMessage(messageBuffer);
+                            if (buff != null && buff.Length > 0)
+                            {
+                                var output = _lowerOutputChannel.Alloc(buff.Length);
+                                output.Write(buff);
+                                await output.FlushAsync();
+                            }
+                            if (securityContext.ReadyToSend)
+                            {
+                                StartWriting(securityContext);
+                            }
+                        }
+                        else if (f == TlsFrameType.Invalid)
+                        {
+                            throw new InvalidOperationException("We have recieved an invalid tls frame");
+                        }
+                        else if (f == TlsFrameType.AppData)
+                        {
+                            var messageBuffer = buffer.Slice(0, pointToSliceMessage);
+                            buffer = buffer.Slice(pointToSliceMessage);
+                            var decryptedData = _outputChannel.Alloc();
+
+                            securityContext.Decrypt(messageBuffer, decryptedData);
+
+                            await decryptedData.FlushAsync();
+                        }
+                        f = buffer.CheckForFrameType(out pointToSliceMessage);
+                    }
+                    _lowerInputChannel.Advance(buffer.Start, buffer.End);
+                }
+            }
+            finally
+            {
+                securityContext.Dispose();
+            }
+        }
+
+        private async void StartWriting<T>(T securityContext) where T : ISecureContext
+        {
+            int maxBlockSize = (SecurityContext.BlockSize - securityContext.HeaderSize - securityContext.TrailerSize);
+            while (true)
+            {
+                var buffer = await _inputChannel.ReadAsync();
+                while (buffer.Length > 0)
+                {
```

Comment:
See my other comment about the read loop. It should look like the other sample I pointed to